### DISCUSSION TITLE
How to predict when using polynomials

### DISCUSSION THREAD

**Discussion by silasprincipe** [inlabru-org/inlabru#151] [Q&A] [answered]
Dear Finn, hi! I'm trying to fit standard "glm" with polynomials through `inlabru` (jut for comparing results for a specific study case), but I'm failling to get predictions with the expected values. To exemplify, I'm running the first example of [the 'Bayesian inference with INLA' book](https://becarioprecario.bitbucket.io/inla-gitbook/ch-smoothing.html):

Everything works fine with INLA:

```
library(inlabru)
library(INLA)
library(SemiPar)
data(lidar)

xx <- as.integer(seq(390, 720, by = 5))
# Add data for prediction
new.data <- cbind(range = xx, logratio = NA)
new.data <- rbind(lidar, new.data)

m.poly <- inla(logratio ~ 1 + range +  I(range^2) + I(range^3),
               data = new.data, control.predictor = list(compute = TRUE, link = 1))

tab <- data.frame(x = xx,
  y = m.poly$summary.fitted[-c(1:nrow(lidar)), "mean"],
  ll95 = m.poly$summary.fitted[-c(1:nrow(lidar)), "0.025quant"],
  ul95 = m.poly$summary.fitted[-c(1:nrow(lidar)), "0.975quant"])

ggplot(aes(x = range, y = logratio), data = lidar) + geom_point() +
  geom_line(aes(x = x, y = y), data = tab, linetype = 1) +
  geom_ribbon(aes(x = x, y = y, ymin = ll95, ymax = ul95), alpha = 0.75, data = tab)

summary(m.poly)

#Fixed effects:
#               mean    sd 0.025quant 0.5quant 0.975quant    mode kld
#(Intercept) -13.443 1.554    -16.494  -13.443    -10.393 -13.443   0
#range         0.074 0.009      0.057    0.074      0.091   0.074   0
#I(range^2)    0.000 0.000      0.000    0.000      0.000   0.000   0
#I(range^3)    0.000 0.000      0.000    0.000      0.000   0.000   0
```

When I try the same with inlabru I'm able to get the same estimates for the effects. However, when I try to use the `predict` function I get very awkward values. I'm probably missing something very obvious, so I apologize in advance.

```
comp <- logratio ~ lin(range) + quad(range^2) + third(range^3)

m.poly <- bru(comp,
             data = lidar,
             family = "gaussian")

summary(m.poly)

#Fixed effects:
#             mean    sd 0.025quant 0.5quant 0.975quant    mode kld
#lin         0.074 0.009      0.057    0.074      0.091   0.074   0
#quad        0.000 0.001     -0.002    0.000      0.002   0.000   0
#third          0.000 0.001     -0.002    0.000      0.002   0.000   0
#Intercept -13.411 1.552    -16.457  -13.411    -10.364 -13.411   0

# Now try to predict:

pred <- predict(m.poly,
                data = data.frame(range = xx), # xx is the object containing the range sequence
                formula = ~ lin + quad + third + Intercept)

head(pred)

# range     mean       sd     q0.025   median   q0.975       smin     smax       cv        var
#1   390 7534.920 53764.51  -85183.81 6463.790 118022.2  -92174.84 176169.2 7.135379 2890622491
#2   395 7828.496 55859.14  -88503.45 6718.345 122618.6  -95765.88 183033.0 7.135361 3120244003
#3   400 8129.598 58007.49  -91908.23 6979.493 127332.9  -99449.01 190072.9 7.135346 3364868723
#4   405 8438.319 60210.22  -95399.25 7247.318 132166.5 -103225.38 197290.9 7.135334 3625270843

### The mean should be something around -0.1 for the lower range
```

Thanks for the help!

**Comment by finnlindgren** [comment, marked as answer]
Sorry for missing this question.

I would normally have guessed there's some numerical issue, since polynomilas and large input values tend to be highly problematic, but in this case it seems that bothe the INLA and bru parameter estimates are reasonable, and that it's just the `predict()` output that is odd.
I'll try to run your code here and check.

**Reply by finnlindgren** [reply]
It turns out that it is indeed a numerical problem that's triggered by the extreme size of the cubic term; `range^3` is in the order `10^8`, which is large enough to make the model very sensitive to minor differences in priors and also purely numerical precision details.  By using `inla.mode="experimental"`, both the `inla()` and `bru()` results improve and become similar.  I didn't have time to test precisely why the original `bru()` results seem to be more affected by the problem than `inla()`, since it's the the same estimation code underneath; the only issue I could think of was that the default prior precision for the intercept is essentially 0 in plain inla, but for inlabru it's instead the same as the default precision for other components. However, setting `+Intercept(1,prec.linear=0)` didn't make a difference, so there must be something else going on as well.

Plain $x^k$ representation of polynomials should generally be avoided unless you can guarantee that the $x$ values are resonably scaled and centred. I'm actually a bit surprised that approach is used for that example in the book, as the $x$ values cause the problem to be extremely badly numerically scaled.

**Reply by finnlindgren** [reply]
Another way to improve the results is to scale the input data in the component inputs:
`~ lin(range/100) + quad((range/100)^2) + third((range/100)^3)`
